File: 3090-maximum-length-substring-with-two-occurrences/3090-maximum-length-substring-with-two-occurrences.java
class Solution {
    public int maximumLengthSubstring(String s) {
        
        // we are using frequency array as constraints are less and only lower case English alphabets 
        int[] freq = new int[26];

        // the question have "longest substring" keywords that suggest the sliding window because substring is a continuous window inside the string

        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {

            int current = s.charAt(right) - 'a'; // get index of current character at right
            freq[current]++; // store occurrence of that character

            // check occurrence of current character is more than 2 times or not, if more than 2 times then shrink window from the left
            while (freq[current] > 2) {
                int removed = s.charAt(left) - 'a'; // get the index of character at left
                freq[removed]--; // remove character at left
                left++; // shrink the window from left
            }

            // update maxlength
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

File: 3090-maximum-length-substring-with-two-occurrences/3090-maximum-length-substring-with-two-occurrences.py
class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        # Frequency array for 26 lowercase English letters
        freq = [0] * 26

        left = 0
        maxLength = 0

        for right in range(len(s)):
            current = ord(s[right]) - ord('a')  # get index of current character at right
            freq[current] += 1  # store occurrence of that character

            # check occurrence of current character is more than 2 times or not
            while freq[current] > 2:
                removed = ord(s[left]) - ord('a')  # get the index of character at left
                freq[removed] -= 1  # remove character at left
                left += 1  # shrink the window from left

            # update maxlength
            maxLength = max(maxLength, right - left + 1)

        return maxLength
